### ISSUE TITLE
Loss becomes NaN during training

### ISSUE THREAD

**Issue by liyizhep** [zwl666666/enfold#3]
When training EnFold on my custom dataset, the loss turns to NaN starting from the second step. Has anyone encountered a similar issue during training, and what areas or steps should I check to troubleshoot this?

[2026-08-12 07:27:16][enfold.models.action_dit][INFO] - ActionDiT backbone payload policy: {'skip_prefixes': ['action_encoder.', 'head.'], 'alpha_scaling': True, 'depth_interpolation': 'linear_align_endpoints', 'tensor_resize': 'sequential_1d_linear_align_corners_true', 'teacher_backend': 'cosmos_predict2.5', 'teacher_num_layers': 28, 'teacher_num_heads': 16, 'teacher_head_dim': 128, 'non_isomorphic_init': {'random_kept_prefixes': ['text_embedding.', 'time_embedding.'], 'zero_filled_suffixes': ['*.bias', 'blocks.*.modulation', 'time_projection.1.bias'], 'shared_time_projection_from_teacher': True}}
[2026-08-12 07:27:16][enfold.models.action_dit][INFO] - Loaded ActionDiT backbone from /data/modelRepository/i2v/i2v_cosmos-predict-2.5/fruit/action_dit_cosmos_init_fruit.pt (keys=160; random_kept_prefixes=['action_encoder.', 'head.']).
[2026-08-12 07:27:18][root][INFO] - cc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -fPIC -c /tmp/tmp5zb6e0rp/test.c -o /tmp/tmp5zb6e0rp/test.o
[2026-08-12 07:27:18][root][INFO] - cc -pthread /tmp/tmp5zb6e0rp/test.o -laio -o /tmp/tmp5zb6e0rp/a.out
[2026-08-12 07:27:19][root][INFO] - cc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -fPIC -c /tmp/tmpiz3u2yn7/test.c -o /tmp/tmpiz3u2yn7/test.o
[2026-08-12 07:27:19][root][INFO] - cc -pthread /tmp/tmpiz3u2yn7/test.o -L/usr/local/cuda -L/usr/local/cuda/lib64 -lcufile -o /tmp/tmpiz3u2yn7/a.out
[2026-08-12 07:27:19][root][INFO] - cc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -fPIC -c /tmp/tmpm9by0mb3/test.c -o /tmp/tmpm9by0mb3/test.o
[2026-08-12 07:27:19][root][INFO] - cc -pthread /tmp/tmpm9by0mb3/test.o -laio -o /tmp/tmpm9by0mb3/a.out
[2026-08-12 07:27:20][megatron.core.msc_utils][INFO] - The multistorageclient package is available.
[2026-08-12 07:27:23][megatron.core.distributed.fsdp.src.megatron_fsdp.megatron_fsdp][INFO] - Detected Megatron Core, using Megatron-FSDP with Megatron.
[2026-08-12 07:27:23][megatron.core.distributed.fsdp.src.megatron_fsdp.param_and_grad_buffer][INFO] - Detected Megatron Core, using Megatron-FSDP with Megatron.
[2026-08-12 07:28:05][enfold.models.cosmos.enfold][INFO] - Reinitialized Cosmos teacher crossattn_proj for mean_pooling input dim=3584.
[2026-08-12 07:28:07][datasets][INFO] - PyTorch version 2.7.0+cu128 available.
[2026-08-12 07:28:09][root][WARNING] - 'torchcodec' is not available in your platform, falling back to 'pyav' as a default decoder
[2026-08-12 07:29:02][enfold.datasets.lerobot.robot_video_dataset][INFO] - Calculating dataset stats for normalization...
[2026-08-12 07:29:43][enfold.runtime][INFO] - Building validation dataset with stats: /data/modelRepository/i2v/i2v_cosmos-predict-2.5/enfold_fruit/2026-08-12_07-26-35/dataset_stats.json
[2026-08-12 07:29:45][root][WARNING] - 'torchcodec' is not available in your platform, falling back to 'pyav' as a default decoder
[2026-08-12 07:29:48][enfold.datasets.lerobot.robot_video_dataset][INFO] - Using dataset stats: /data/modelRepository/i2v/i2v_cosmos-predict-2.5/enfold_fruit/2026-08-12_07-26-35/dataset_stats.json
[2026-08-12 07:29:48][enfold.trainer][INFO] - Accelerate training: distributed_type=DistributedType.DEEPSPEED zero_stage=1 world_size=1 process_index=0 cfg_mixed_precision=bf16 accelerator_mixed_precision=bf16 grad_accum=2 grad_clip=1.0000
[2026-08-12 07:29:48][enfold.trainer][INFO] - using accelerator.device=cuda:0
[2026-08-12 07:29:56][enfold.trainer][INFO] - Train/val dataset size: 740291/7593
[2026-08-12 07:29:56][enfold.trainer][INFO] - Setting DiT to train mode and freezing other model components.
[2026-08-12 07:29:56][enfold.trainer][INFO] - Starting training with max_steps=100000.
[2026-08-12 07:30:09][enfold.trainer][INFO] - [train] epoch=0 step=1/100000 loss=2.5691 loss_action=1.8734 loss_repr=0.3780 loss_video=0.3178 lr=4.00e-08 speed=0.08 step/s, 0.31 samples/s eta=358:46:11
[2026-08-12 07:30:14][enfold.trainer][INFO] - [train] epoch=0 step=2/100000 loss=nan loss_action=2.4608 loss_repr=nan loss_video=nan lr=6.00e-08 speed=0.11 step/s, 0.45 samples/s eta=247:52:41
[2026-08-12 07:30:19][enfold.trainer][INFO] - [train] epoch=0 step=3/100000 loss=nan loss_action=nan loss_repr=nan loss_video=nan lr=8.00e-08 speed=0.13 step/s, 0.51 samples/s eta=218:00:51
[2026-08-12 07:30:24][enfold.trainer][INFO] - [train] epoch=0 step=4/100000 loss=nan loss_action=nan loss_repr=nan loss_video=nan lr=1.00e-07 speed=0.14 step/s, 0.57 samples/s eta=196:14:40
[2026-08-12 07:30:29][enfold.trainer][INFO] - [train] epoch=0 step=5/100000 loss=nan loss_action=nan loss_repr=nan loss_video=nan lr=1.20e-07 speed=0.15 step/s, 0.60 samples/s eta=184:44:47
[2026-08-12 07:30:34][enfold.trainer][INFO] - [train] epoch=0 step=6/100000 loss=nan loss_action=nan loss_repr=nan loss_video=nan lr=1.40e-07 speed=0.16 step/s, 0.63 samples/s eta=176:34:26
[2026-08-12 07:30:39][enfold.trainer][INFO] - [train] epoch=0 step=7/100000 loss=nan loss_action=nan loss_repr=nan loss_video=nan lr=1.60e-07 speed=0.16 step/s, 0.65 samples/s eta=170:52:41
[2026-08-12 07:30:44][enfold.trainer][INFO] - [train] epoch=0 step=8/100000 loss=nan loss_action=nan loss_repr=nan loss_video=nan lr=1.80e-07 speed=0.17 step/s, 0.66 samples/s eta=168:07:26
[2026-08-12 07:30:47][enfold.trainer][INFO] - [train] epoch=0 step=9/100000 loss=nan loss_action=nan loss_repr=nan loss_video=nan lr=2.00e-07 speed=0.17 step/s, 0.70 samples/s eta=159:10:58
[2026-08-12 07:30:52][enfold.trainer][INFO] - [train] epoch=0 step=10/100000 loss=nan loss_action=nan loss_repr=nan loss_video=nan lr=2.20e-07 speed=0.18 step/s, 0.71 samples/s eta=157:12:27

**Comment by xingyitong1** [comment]
Thanks for reporting this. We haven't been able to identify the cause from the training log alone.

Since the NaN first appears in loss_repr and loss_video at step 2, could you please provide a bit more information about your setup?

In particular, it would be helpful if you could share:
- The full training command you used, including all Hydra overrides.
- The resolved training configuration, especially the data, model, and task configs.
- Your custom dataset configuration, including:  action / proprio dimensions, observation and video shapes, FPS / temporal sampling settings, normalization settings
- The generated dataset_stats.json. 

Please also check whether any statistics contain NaN, Inf, zero/near-zero std, or unusually large values.

It would also be very useful to know whether training on the official LIBERO or RoboTwin dataset works correctly in the same environment, and whether the NaN only occurs with your custom dataset.

Also, please note that the default hyperparameters were tuned for multi-GPU training. If your global batch size is much smaller, consider reducing the learning rate according to linear scaling, or increasing gradient accumulation steps to achieve a larger effective global batch size.

Thanks!

**Comment by liyizhep** [comment]
Thank you for the detailed response, it was really helpful!

It turns out that the NaN issue was indeed caused by an inappropriate batch size. Yesterday, I was debugging on a single GPU with a global batch size of 8. Today, I tested with 2 GPUs and gradient accumulation (global batch size = 32), and the NaN issue completely disappeared. 

Thanks again, and I'm really looking forward to seeing Enfold work effectively in our own scenario!

**Comment by zwl666666** [comment]
> Thank you for the detailed response, it was really helpful!
> 
> It turns out that the NaN issue was indeed caused by an inappropriate batch size. Yesterday, I was debugging on a single GPU with a global batch size of 8. Today, I tested with 2 GPUs and gradient accumulation (global batch size = 32), and the NaN issue completely disappeared.
> 
> Thanks again, and I'm really looking forward to seeing Enfold work effectively in our own scenario!

Glad to hear that the issue has been resolved, and thanks for sharing the finding! This is also very useful feedback for us.

For the detailed training hyperparameters and policy settings, we would recommend aligning them as closely as possible with Table 6: Training and policy configuration in the paper. This should provide a more stable starting point before tuning the configuration for your specific setup.

We’re excited to hear that you’re trying EnFold in your own scenario, and we’d be very interested to hear how it performs. Thanks again for trying EnFold!

**Comment by liyizhep** [comment]
Over the past few weeks, we tested simple pick-and-place task and Enfold achieved high success rates. However, our training process currently feels somewhat inefficient, so we would like to ask a few questions regarding your training protocol and experience:

1. **Dataset size for positional generalization**: In the paper, you mentioned training a multi-task model with 400 trajectories in total (100 trajectories per task across 4 tasks). If we want to train a pick-and-place task with spatial/positional generalization (i.e., successfully grasping the target object placed anywhere within a designated region), would 100 expert demonstrations with positional diversity be sufficient?
2. **Action loss convergence**: In the paper, models are typically trained for around 10 epochs. Based on your experience, what target range should the action loss converge to in order to achieve reliable real-robot success rates? Is a loss level of 0.01–0.02 low enough?

Thank you for your guidance and help!

**Comment by xingyitong1** [comment]
Hi, we are very happy to hear that Enfold has achieved high success rates in your pick-and-place setting. Thank you very much for sharing your experience and for the thoughtful questions.
1. For positional diversity, 100 demonstrations can support generalization within the approximate workspace near the robot arms. Generalizing across the full designated region, especially to positions farther from the arms, will likely require more data. We have not yet systematically studied the optimal data allocation for this setting, so we cannot give a precise recommendation yet. We would be happy to exchange observations as you experiment with this setting. For multi-task training, it is also important to align the data pipeline and robot setup across tasks; otherwise, differences between tasks may lead to interference.
2. In our experience, an action loss converging to around 0.01 is generally sufficient for reliable performance. However, the appropriate threshold depends on the task. Some simpler tasks may work well before the loss reaches 0.01. We recommend saving multiple checkpoints during training and evaluating them on the real robot to select the best one.

We hope this is helpful. Please feel free to continue the discussion if you have further questions or observations to share.
